### ISSUE TITLE
Use gradle cache to skip execution

### ISSUE THREAD

**Issue by victorpineda-jt** [takahirom/roborazzi#247]
After migrating from Paparazzi to Roborazzi, I noticed a different behaviour when using gradle's cache between both libraries. While Paparazzi is skipping all the recording steps if the tests that have not changed and the screenshots already exist, Roborazzi records new screenshots every time the record command is executed, no matter if the test has changed or not. That leads to super quick execution times in Paparazzi when executing the command iteratively compared to Roborazzi. For example, if I execute the record command twice with Paparazzi, the second time it takes around 5s, while in Roborazzi, it takes around 1 minute. 

Is there any way to change the caching behaviour to behave as Paparazzi?

I've already tried adding in my build.gradle.kts file the option suggested in the doc:

```
roborazzi {
    outputDir = "src/your/screenshot/folder"
}
```

In my case, it's a kts file so I'm setting it like this:

```
roborazzi {
    outputDir.set(File("./src/test/cache"))
}
```

The `cache` folder is created but images are still recorded in the default output dir `/build/outputs/roborazzi`. The task is never executed from the cache:

<img width="507" alt="Captura de pantalla 2024-01-18 a las 10 35 36" src="https://github.com/takahirom/roborazzi/assets/155441962/e13d2279-5cea-4e49-9bb0-195bd1bcdfec">

I'm using the latest version 1.8.0-rc2.

Thanks in advance for the library and all your effort!

**Comment by victorpineda-jt** [comment]
@takahirom this is what I get when using --info:

```
> Task :feature:login:recordRoborazziStagingDebug
Skipping task ':feature:login:recordRoborazziStagingDebug' as it has no actions.

> Task :feature:login:finalizeTestRoborazziStagingDebug
Roborazzi: roborazziTestFinalizer.onlyIf doesRoborazziRun true
Caching disabled for task ':feature:login:finalizeTestRoborazziStagingDebug' because:
  Gradle would require more information to cache this task
Task ':feature:login:finalizeTestRoborazziStagingDebug' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Roborazzi: roborazziTestFinalizer.doLast
Copy files from /Users/victor.pineda@jobandtalent.com/Projects/android-companies/feature/login/src/test/cache to /Users/victor.pineda@jobandtalent.com/Projects/android-companies/feature/login/build/intermediates/roborazzi
Save result to /Users/victor.pineda@jobandtalent.com/Projects/android-companies/feature/login/build/test-results/roborazzi/results-summary.json with results:21

```

**Comment by takahirom** [comment]
@victorpineda-jt 
I'm sorry, I need information about the Skip for :testDebugUnitTest.

**Comment by victorpineda-jt** [comment]
Sure @takahirom , when running `feature:login:testStagingDebugUnitTest --info` I get:

```
Build cache key for task ':feature:login:testStagingDebugUnitTest' is 15840abd333c9ec532196b965e3dc5a2
Skipping task ':feature:login:testStagingDebugUnitTest' as it is up-to-date.
Resolve mutations for :feature:login:finalizeTestRoborazziStagingDebug (Thread[Execution worker Thread 8,5,main]) started.
:feature:login:finalizeTestRoborazziStagingDebug (Thread[Execution worker Thread 8,5,main]) started.
> Task :feature:login:finalizeTestRoborazziStagingDebug SKIPPED
Roborazzi: roborazziTestFinalizer.onlyIf doesRoborazziRun false
Skipping task ':feature:login:finalizeTestRoborazziStagingDebug' as task onlyIf 'Task satisfies onlyIf spec' is false.
BUILD SUCCESSFUL in 705ms
```

In this case, the test task is being skipped.

**Comment by takahirom** [comment]
To understand why it was not skipped, it would be easier to find out if there are logs from when it was not skipped

**Comment by victorpineda-jt** [comment]
@takahirom are you referring to the task` : recordRoborazziDebug --info` ?

I've searched for more logs about the task `testDebugUnitTest` when running that command. 

```
Build cache key for task ':feature:login:transformStagingDebugUnitTestClassesWithAsm' is 7655558f9710a005795ba57182f8a3f6
Skipping task ':feature:login:transformStagingDebugUnitTestClassesWithAsm' as it is up-to-date.
Resolve mutations for :feature:login:testStagingDebugUnitTest (Thread[Execution worker Thread 5,5,main]) started.
:feature:login:testStagingDebugUnitTest (Thread[Execution worker Thread 5,5,main]) started.
Gradle Test Executor 221 started executing tests.

> Task :feature:login:testStagingDebugUnitTest
Roborazzi: test.outputs.upToDateWhen !(doesRoborazziRun:true && inputDirEmpty:true)
Transforming kotlin-parcelize-runtime-1.9.22.jar (org.jetbrains.kotlin:kotlin-parcelize-runtime:1.9.22) with IdentityTransform
.
. // I've deleted a bunch of transforming logs here
.
Transforming android.jar with MockableJarTransform
Build cache key for task ':feature:login:testStagingDebugUnitTest' is 2f757cafa5338e1ba2be5269ae294507
Task ':feature:login:testStagingDebugUnitTest' is not up-to-date because:
  Task.upToDateWhen is false.
Roborazzi: test.doFirst
Transforming android.jar with MockableJarTransform
Transforming android.jar with MockableJarTransform
Starting process 'Gradle Test Executor 221'. Working directory: /Users/victor.pineda/Projects/android-companies/feature/login Command: /Library/Java/JavaVirtualMachines/zulu-17.jdk/Contents/Home/bin/java -Dcoverage.ignore.private.constructor.util.class=true -Didea.coverage.log.level=error -Didea.new.sampling.coverage=true -Didea.new.tracing.coverage=true -Djava.awt.headless=true -Dorg.gradle.internal.worker.tmpdir=/Users/victor.pineda/Projects/android-companies/feature/login/build/tmp/testStagingDebugUnitTest/work -Dorg.gradle.native=false -Droborazzi.test.compare=false -Droborazzi.test.record=true -Droborazzi.test.verify=false -javaagent:/Users/victor.pineda/.gradle/caches/modules-2/files-2.1/org.jetbrains.intellij.deps/intellij-coverage-agent/1.0.683/e36e51c64ffeac98be4e5ea807fc963327f424dc/intellij-coverage-agent-1.0.683.jar=/Users/victor.pineda/Projects/android-companies/feature/login/build/tmp/testStagingDebugUnitTest/intellijagent.args @/Users/victor.pineda/.gradle/.tmp/gradle-worker-classpath8076781847995570035txt -Xmx512m -Dfile.encoding=UTF-8 -Duser.country=ES -Duser.language=es -Duser.variant -ea worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 221'
Successfully started process 'Gradle Test Executor 221'

OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
Gradle Test Executor 221 finished executing tests.

> Task :feature:login:testStagingDebugUnitTest
Finished generating test XML results (0.012 secs) into: /Users/victor.pineda/Projects/android-companies/feature/login/build/test-results/testStagingDebugUnitTest
Generating HTML test report...
Finished generating test html results (0.008 secs) into: /Users/victor.pineda/Projects/android-companies/feature/login/build/reports/tests/testStagingDebugUnitTest
Stored cache entry for task ':feature:login:testStagingDebugUnitTest' with cache key 2f757cafa5338e1ba2be5269ae294507
Resolve mutations for :feature:login:finalizeTestRoborazziStagingDebug (Thread[Execution worker Thread 5,5,main]) started.
Resolve mutations for :feature:login:recordRoborazziStagingDebug (Thread[Execution worker Thread 4,5,main]) started.
:feature:login:recordRoborazziStagingDebug (Thread[Execution worker Thread 4,5,main]) started.
:feature:login:finalizeTestRoborazziStagingDebug (Thread[Execution worker Thread 5,5,main]) started.

> Task :feature:login:recordRoborazziStagingDebug
Skipping task ':feature:login:recordRoborazziStagingDebug' as it has no actions.

> Task :feature:login:finalizeTestRoborazziStagingDebug
Roborazzi: roborazziTestFinalizer.onlyIf doesRoborazziRun true
Caching disabled for task ':feature:login:finalizeTestRoborazziStagingDebug' because:
  Gradle would require more information to cache this task
Task ':feature:login:finalizeTestRoborazziStagingDebug' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Roborazzi: roborazziTestFinalizer.doLast
Copy files from /Users/victor.pineda/Projects/android-companies/feature/login/src/test/cache to /Users/victor.pineda/Projects/android-companies/feature/login/build/intermediates/roborazzi
Save result to /Users/victor.pineda/Projects/android-companies/feature/login/build/test-results/roborazzi/results-summary.json with results:21

0 problems were found storing the configuration cache.

See the complete report at file:///Users/victor.pineda/Projects/android-companies/build/reports/configuration-cache/ahw653op2l8f148k97i1gwe54/1n35nftlbfevbzd2rdfnjou7q/configuration-cache-report.html

BUILD SUCCESSFUL in 24s
417 actionable tasks: 25 executed, 392 up-to-date
Watched directory hierarchies: [/Users/victor.pineda/Projects/android-companies]
Configuration cache entry stored.

Build Analyzer results available
12:31:11: Execution finished 'feature:login:recordRoborazziStagingDebug --info'.
```

**Comment by takahirom** [comment]
Thanks! I'll take a loot at this!

**Comment by takahirom** [comment]
I suspect the cause might be "inputDirEmpty:true", but are there images in the output folder when you test? I do believe they exist, but I just want to confirm.

**Comment by victorpineda-jt** [comment]
yes, all the images exists in the default output dir /build/outputs/roborazzi. When I noticed this cache issue, I also added in the build.gradle.kts file the configuration mentioned above:

```
roborazzi {
    outputDir.set(File("./src/test/cache"))
}
```

But this output dir is being ignored and it is empty. Images are still stored in the same directory. This configuration in the `build.gradle.kts` file does not affect the caching. By the way, I'm not using any roborazzi test rule to change the output dir at this moment.

**Comment by takahirom** [comment]
But, yes, I definitely feel that it should be possible to specify this from Gradle.

**Comment by victorpineda-jt** [comment]
@takahirom Thanks for the clarification. I keep investigating this and seems tricky. After executing this command again, I've noticed that it now displays it as UP-TO-DATE:

```
> Task :feature:login:testStagingDebugUnitTest UP-TO-DATE
Roborazzi: test.outputs.upToDateWhen !(doesRoborazziRun:true && inputDirEmpty:false)
```

The command takes only 1 second to run instead of 24s, so it looks like the cache is working fine for this specific module. 
Next, I've executed the same command for another module. This other module already has the screenshots in the corresponding output dir /build/outputs/roborazzi. This other module also has the same build.gradle.kts config:

```
roborazzi {
    outputDir.set(File("./src/test/cache"))
}

```

However, after executing the record command for the new module in multiple occasions (`feature:workers:recordRoborazziStagingDebug --info`), I get always the same result:

```
> Task :feature:workers:testStagingDebugUnitTest
Roborazzi: test.outputs.upToDateWhen !(doesRoborazziRun:true && inputDirEmpty:true)
```

The output images are created again for this other module. 

If I go back and execute the same command for the previous module (`login`), the command is still caching the task and not updating the output images.

I've compared both `build.gradle.kts` files and they are almost identical so not sure why in the second module, the task is not being cached.

**Comment by victorpineda-jt** [comment]
@takahirom I think I've found the root issue. The output dir set in the build.gradle.kts file was not matching the output dir where the screenshots were stored. After changing it to match the output directory for screenshots, the tasks are cached as expected. What confused me was that it was working fine in the first module where I tried it (`login`), but not in others. Then I noticed that the output dir declared in the build.gradle.kts files actually contained screenshots for the `login` module, not sure why, as I didn't configure any rule to use that output directory. Anyway, after updating the directory, everything seems to be working fine! Thanks for all your help!
